src/js/common/hashString.js

```diff
@@ -0,0 +1,63 @@
+(function (fluid) {
+    "use strict";
+
+    var gpii = fluid.registerNamespace("gpii");
+    fluid.registerNamespace("gpii.schema");
+
+    /**
+     *
+     * Break down a string into 50,000 character blocks and construct a composite "hash" based on each "segment".  The
+     * smaller segment size is meant to avoid the risk of "collisions" when comparing very large strings (the previous
+     * approach collided infrequently but consistently when comparing 175,000 close variations on a 175,000 character
+     * string).
+     *
+     * Strings shorter than 50,000 characters will return a hash that looks like a single stringified 32-bit integer, as
+     * in:
+     *
+     * `-1864109329`
+     *
+     * Strings longer than 50,000 characters will return a hash that looks like multiple stringified 32-bit integers
+     * joined using a colon, as in:
+     *
+     * `1464911337:-123853895:59841156:-143877442`
+     *
+     * @param {String} originalString - The original string to be hashed.
+     * @return {String} - A hash composed of the hashes for each sub-segment in the original string.
+     *
+     */
+    gpii.schema.hashString = function (originalString) {
```

Comment:
Whilst you've worked hard to reduce the probability of a hash collision, I feel that in practice we simply can't afford any - at least, "essentially any". The following table explains that, for example, if we are just considering the space of strings shorter than 50k which just get a single 32-bit hash, after only 100,000 strings the chance of a collision becomes greater than 1 in 2. https://stackoverflow.com/questions/14210298/probability-of-collision-when-using-a-32-bit-hash
In any case, the "correct" data structure to be used (either an open or a closed hashtable https://stackoverflow.com/questions/9124331/meaning-of-open-hashing-and-closed-hashing ) would unavoidably store the actual key of the hash somewhere, in order to detect whether a collision has occurred.
In this case I recommend we simply store the key itself - either the JSON.stringify version of the schema or a lightly compacted version thereof, and thus eliminate all the complexity due to this hashing. We can revisit this in future if it becomes a performance worry - but bear in mind that we can't avoid storing the schema itself in memory, and this will at most double that storage cost.

Reply:
I'm happy to get rid of it and review if we find that it would be useful to have a shorter hash in the future.